controllers/front/SmailyCustomerCron.php

```diff
@@ -44,28 +54,21 @@ private function syncContacts()
     {
         if (Configuration::get('SMAILY_ENABLE_CRON') === "1") {
```

Comment:
This is an inefficient way of synchronizing customers. Problems that I see are:

- All unsubscribed subscribers are loaded into memory;
- Fetching existing customers is not required to check if unsubscribeable email address exist in the database;
- Each customer is updated individually;
- Configuration key is accessed in each loop iteration;

What you can do is use a pattern like:
```php
$offset = 0;
while (true) {
    $unsubscribers = $this->module->callApi('contact', array(
        'list' => 2,
        'limit' => self::UNSUBSCRIBERS_BATCH_LIMIT,
        'offset' => $offset,
    ));

    if (empty($unsubscribers)) {
        break;
    }

    $query = 'UPDATE ' . _DB_PREFIX_ . 'customer SET newsletter=0 WHERE email IN (' . implode(',', array_map(function($item) {
        return '"' . pSQL($item['email']) . '"';
    }, $unsubscribers));
    Db::getInstance()->execute($query);

    $offset++;
}
```

The benefit of using `UPDATE` query is that it takes care of not knowing if the customer already exists or not.

Same logic can (and should) be applied to synchronizing customer data (at the moment it is also loading all of the data into memory).

Reply:
Nice genius query solution, never thought of it. Refactored accordingly https://github.com/sendsmaily/smaily-prestashop-module/pull/10/commits/fa51248d09a0cfbd37f339b0208444869b1f62ac